File: client_mobile/app/src/main/java/com/example/tiktokapp/services/PostService.java
package com.example.tiktokapp.services;

import com.example.tiktokapp.responseModel.APIRespone;
import com.example.tiktokapp.responseModel.APIResponeList;
import com.example.tiktokapp.responseModel.FollowAPIRespone;
import com.example.tiktokapp.responseModel.Post;
import com.example.tiktokapp.responseModel.SimpleAPIRespone;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.Part;
import retrofit2.http.Path;

public interface PostService {

    @GET("post")
    Call<APIResponeList<Post>> getPosts();

    @POST("post/like/{postId}")
    Call<SimpleAPIRespone> likePost(@Path("postId") int postId);

    @GET("post/user/{userId}")
    Call<APIResponeList<Post>> getPostsByUserId(@Path("userId") int userId);

    @POST("post/unlike/{postId}")
    Call<SimpleAPIRespone> unlikePost(@Path("postId") int postId);

    @Multipart
    @POST("post/upload")
    Call<APIRespone<Post>> uploadPost(@Part("title") RequestBody title,
                                      @Part("visibility") RequestBody visibility,
                                      @Part MultipartBody.Part video
                                      );

    @Multipart
    @POST("post/upload")
    Call<APIRespone<Post>> uploadPostWithThumbnail(
                                        @Part("title") RequestBody title,
                                        @Part("visibility") RequestBody visibility,
                                        @Part MultipartBody.Part video,
                                        @Part MultipartBody.Part thumnail
    );

    @GET("post/{postId}")
    Call<APIRespone<Post>> getPostById(@Path("postId") int postId);
}

File: client/src/features/post/postService.ts
import { axiosNoToken, axiosToken } from '@/axios';
import { PaginationModel } from '@/models';
import { PostModel } from '@/models/post';
import AbstractPayload from '@/utils/abtractPayloadType';
const routePath = '/post';

export interface PostsPayload extends AbstractPayload {
  posts: PostModel[];
  pagination: PaginationModel;
}
export interface PostPayload extends AbstractPayload {
  post: PostModel;
}
const PostService = {
  async getPosts(title?: string) {
    return new Promise<PostsPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.get(
          routePath + `/${title ? `?title=${title}` : ''}`
        );
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async watchPost(postId?: number) {
    return new Promise<PostsPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosNoToken.post(routePath + `/${postId}/watch`);
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async getPostsByUserId(userId: number) {
    return new Promise<PostsPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.get(routePath + `/user/${userId}`);
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async getFollowingPosts() {
    return new Promise<PostsPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.get(routePath + '/following');
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async getFriendPosts() {
    return new Promise<PostsPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.get(routePath + '/friends');
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async getPostById(postId: number) {
    return new Promise<PostPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.get(routePath + `/${postId}`);
        const data = resp.data;
        const transformedPost = {
          ...data.post,
          isMe: Boolean(data.post.isMe),
          isFollow: Boolean(data.post.isFollow),
          isFriend: Boolean(data.post.isFriend),
          isLiked: Boolean(data.post.isLiked),
        };

        resolve({ ...data, post: transformedPost });
      } catch (error) {
        reject(error);
      }
    });
  },

  async sharePost(postId: number) {
    return new Promise<AbstractPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.post(routePath + '/share/' + postId);
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async likePost(postId: number) {
    return new Promise<AbstractPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.post(routePath + '/like/' + postId);
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async unlikePost(postId: number) {
    return new Promise<AbstractPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.post(routePath + '/unlike/' + postId);
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
  async uploadPost(postFormData: FormData) {
    return new Promise<AbstractPayload>(async (resolve, reject) => {
      try {
        const resp = await axiosToken.post(
          routePath + '/upload/',
          postFormData,
          {
            headers: {
              'Content-Type': 'multipart/form-data',
            },
          }
        );
        resolve(resp.data);
      } catch (error) {
        reject(error);
      }
    });
  },
};
export default PostService;
